app/controllers/user_upgrade_requests_controller.rb

```diff
@@ -0,0 +1,23 @@
+class UserUpgradeRequestsController < ApplicationController
+  def new
+    @user_upgrade_request = UserUpgradeRequest.new
+  end
+
+  def show; end
+
+  def create
+    @user_upgrade_request = UserUpgradeRequest.new(user_upgrade_request_params)
+
+    if @user_upgrade_request.valid?
+      UserUpgradeRequestService.new(current_user).request_upgrade
```

Comment:
I went down this route because the form is so simple and I think the code in the service is likely to be triggered in other ways in the future. Maybe it's a YAGNI thing, but while we don't have a requirement for it yet but it doesn't seem too far fetched that the engagement team might want to trigger an email from the /users screen for example.

I guess either way works, as long as the functionality is kept out of the controller and testable.

I nearly used a form object, like we do elsewhere to keep things consistent. But the form is really so small that it feels hard to justify it. The checkbox isn't really used anywhere else, it's just a guard to ensure the the user has ticked it. Because of this I was tempted to do the form totally in the controller. We only need to make sure the checkbox is true, which comes down to checking a param. The only which stopped me was that I needed the error state, which winds up being more lines of code because I don't get to use the form helpers.

Reply:
Yes, I also tried initially to get away without a form object and also quickly found that the errors are difficult to deal with without it!